### ISSUE TITLE
What are state keys actually *for*?

### ISSUE THREAD

**Issue by joepie91** [matrix-org/matrix-doc#2144]
The client-server spec says (emphasis mine):

> State Events: These are events which update the metadata state of the room (e.g. room topic, room membership etc). __State is keyed by a tuple of event type and a state_key. State in the room with the same key-tuple will be overwritten.__

And:

> state_key: Required. A unique key which defines the overwriting semantics for this piece of room state. This value is often a zero-length string. The presence of this key makes this event a State Event. State keys starting with an @ are reserved for referencing user IDs, such as room members. With the exception of a few events, state events set with a given user's ID as the state key MUST only be set by that user.

... which doesn't really clarify what the *purpose* of a state key is - why was the concept introduced into the protocol?

The second snippet says "defines the overwriting semantics", but this is really vague - what is meant with "overwriting semantics", what is the overarching format of the state key (if any), and how does that relate to the apparently different semantics options? Shouldn't it then be an enum, rather than a free-form string? What are the implications of a state key changing, or being omitted?

Basically, the high-level explanation of "what are state keys, why do they exist, why should you care about them" is missing.

**Comment by t3chguy** [comment]
> ... which doesn't really clarify what the purpose of a state key is - why was the concept introduced into the protocol?


Lets use `m.room.member` (the membership event) as an example,
The state key in the case of this event is the user in question who is represented by this state event, so this offers multiple benefits, one is you can quickly access the state event which tells you that user's displayname and avatar,
It also acts as compartmentalization, only that user can write to that event, if all memberships were stored under keys within one event things would get a lot more confusing in terms of checking who has access to make a certain edit to an event. It also prevents merge conflicts between different servers making changes to the event at the same time, given that only one user/server actually has write permission to that event.

The permissioning is not quite as clear with other events, another example is Widgets, each widget is its own state event with the widget ID as the state key but anyone with a sufficient power level in a room can edit them

I seem to recall state keys which resemble MXIDs are reserved for only that MXID to have write access to, but did not verify that against the spec at this time.

**Comment by t3chguy** [comment]
> The second snippet says "defines the overwriting semantics", but this is really vague - what is meant with "overwriting semantics", what is the overarching format of the state key (if any), and how does that relate to the apparently different semantics options? Shouldn't it then be an enum, rather than a free-form string? What are the implications of a state key changing, or being omitted?

Lets say you wanted to update a membership event and state keys weren't a thing, you would have to update the entire event, re-sending all the details you didn't change (and didn't delete) as well as the changed bits, that's potentially a huge payload and could clash with someone else's changes.
If you send a state event, it only overwrites the previous state event with the same event type and same state key, which makes the overwrites much smaller.

> Shouldn't it then be an enum, rather than a free-form string? What are the implications of a state key changing, or being omitted?

State Keys are non-optional in state events, they default to the empty string, and most state events leave them as such.
The implication of one being omitted or changed would be that the event would exist alongside the other state events with the same event type but different state keys on the current state of the room

**Comment by t3chguy** [comment]
What the state key should be set to is down to the event type, if you were to invent your own type, `pie.joe.91.setting` event for example, you'd define the semantics of what the state_key denotes, if you just wanted one of these settings to apply at any given time then you'd leave the state_key = `""`

Whereas if you wanted to have an event per "THING", e.g per-room if it was an event to control some IoT device in your house then the state key might be the unique name of that room

**Comment by joepie91** [comment]
> Lets say you wanted to update a membership event and state keys weren't a thing, you would have to update the entire event, re-sending all the details you didn't change (and didn't delete) as well as the changed bits, that's potentially a huge payload and could clash with someone else's changes.
If you send a state event, it only overwrites the previous state event with the same event type and same state key, which makes the overwrites much smaller.

Right, I would argue that the phrasing in the spec is wrong or at the very least misleading, then - it *doesn't* define the overwriting semantics (which are always the same, a full overwrite), it just subdivides the data into smaller overwritable parts.

----

Somebody on IRC also pitched in with some clarification, and from that, I derived this explanation:

__The purpose of a 'state key' is to subdivide a given state type into many smaller 'slots' for that state type, which is useful when a state event defines the state for something *in* the room (eg. a user), rather than the room itself. Each such thing has its own slot, that can be individually overridden.__

Does that sound about right?

**Comment by t3chguy** [comment]
It sounds reasonable
